service/src/main/java/org/cbioportal/service/impl/StudyServiceImpl.java

```diff
@@ -33,21 +34,24 @@
                                            String sortBy, String direction) {
 
         List<CancerStudy> allStudies = studyRepository.getAllStudies(keyword, projection, pageSize, pageNumber, sortBy, direction);
+        Map<String,CancerStudy> allStudiesByCancerStudyIdentifier = allStudies.stream().collect(Collectors.toMap(c -> c.getCancerStudyIdentifier(), c -> c));
+
         if (keyword != null && (pageSize == null || allStudies.size() < pageSize)) {
             List<CancerStudy> primarySiteMatchingStudies = findPrimarySiteMatchingStudies(keyword);
-            if (pageSize != null) {
-                int toIndex = primarySiteMatchingStudies.size() > pageSize - allStudies.size() ? 
-                    pageSize - allStudies.size() : primarySiteMatchingStudies.size();
-                primarySiteMatchingStudies = primarySiteMatchingStudies.subList(0, toIndex);
-            }
             for (CancerStudy cancerStudy : primarySiteMatchingStudies) {
-                if (!allStudies.stream().anyMatch(c -> c.getCancerStudyIdentifier().equals(cancerStudy.getCancerStudyIdentifier()))) {
-                    allStudies.add(cancerStudy);
+                if (!allStudiesByCancerStudyIdentifier.containsKey(cancerStudy.getCancerStudyIdentifier())) {
+                    allStudiesByCancerStudyIdentifier.put(cancerStudy.getCancerStudyIdentifier(), cancerStudy);
+                }
+                if (pageSize != null && allStudiesByCancerStudyIdentifier.size() == pageSize) {
```

Comment:
Looks good to me! Not sure if the `pageSize != null` is necessary.

Reply:
i believe so, in case you don't specify pageSize as parameter to the request it will be null here. We want to return all the found matches in that case, so never break 🏎